### PR TITLE
AP-1979: Add new fields to gross income summary with ability to call assessment version 3

### PR DESCRIPTION
**Pull request by Alex-Swann**
## What

[https://dsdmoj.atlassian.net/browse/AP-1979](https://dsdmoj.atlassian.net/browse/AP-1979)

Added new fields for the gross summary income

## Checklist

Before you ask people to review this PR:

- [x] Tests and rubocop should be passing: `bundle exec rake`
- [x] Github should not be reporting conflicts; you should have recently run `git rebase master`.
- [x] There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- [x] The PR description should say what you changed and why, with a link to the JIRA story.
- [x] You should have looked at the diff against master and ensured that nothing unexpected is included in your changes.
- [x] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: app/services/decorators/other_income_source_decorator.rb [hunk 1]
```diff
@@ -4,16 +4,32 @@ def initialize(record)
...
+      end
     end
 
     private
 
+    def find_by(transaction_type:)
+      income_categories = CFEConstants::VALID_INCOME_CATEGORIES.map(&:to_sym)
+      income_categories.reduce({}) { |hash, cat| hash.update(cat => @record["#{cat}_#{transaction_type}"]) }
+    end
+
```
Comment (stephenrichards) [inline app/services/decorators/other_income_source_decorator.rb:32]: I don't like this.  First off, the method name is misleading.  `#find_by` is a common ActiveRecord method used for retrieving a record from the database, so it gives a misleading impression.  Secondly, it's not at all clear what's going on.  I'd much rather see something more verbose, but clear in intention.

File: spec/factories/assessment_factory.rb [hunk 2]
```diff
@@ -4,6 +4,11 @@
     remote_ip { Faker::Internet.ip_v4_address }
     submission_date { Date.current }
     matter_proceeding_type { 'domestic_abuse' }
+    version { '2' }
+
+    trait :with_latest_version do
+      version { '3' }
```
Comment (stephenrichards) [inline spec/factories/assessment_factory.rb:10]: A trait with the name `with_latest_version` is going to become out of date when we introduce version 4 (sad, but inevitible).  Why not just call it `:v3` ?

File: spec/factories/other_income_source_factory.rb [hunk 3]
```diff
@@ -6,6 +6,10 @@
     name { CFEConstants::VALID_INCOME_CATEGORIES.sample }
     monthly_income { nil }
 
+    trait :with_latest_version do
+      gross_income_summary { create :gross_income_summary, :with_latest_version }
```
Comment (stephenrichards) [inline spec/factories/other_income_source_factory.rb:10]: see note above about trait `with_latest_version`

File: spec/factories/assessment_factory.rb (hunk 2 above)
Comment (Alex-Swann) [reply spec/factories/assessment_factory.rb:10]: Yeah I think that crossed my mind too and I forgot to change it 😂

Comment (Alex-Swann) [reply spec/factories/assessment_factory.rb:10]: I think I was contemplating using latest version number from CFEConstants but let's keep it simple and do that 👍

File: app/services/decorators/gross_income_summary_decorator.rb [hunk 4]
```diff
@@ -10,21 +10,36 @@ def initialize(gross_income_summary)
...
+          upper_threshold: record.upper_threshold,
+          assessment_result: record.assessment_result,
+          monthly_income_equivalents: MonthlyIncomeEquivalentDecorator.new(record).as_json,
+          monthly_outgoing_equivalents: MonthlyOutgoingEquivalentDecorator.new(disposable_income_summary).as_json,
+          state_benefits: record.state_benefits.map { |sb| StateBenefitDecorator.new(sb).as_json },
+          other_income: record.other_income_sources.map { |oi| OtherIncomeSourceDecorator.new(oi).as_json },
+          irregular_income: IrregularIncomePaymentsDecorator.new(record.irregular_income_payments).as_json
+        }
+      end
     end
```
[the remark this answers is not in the document]
Comment (Alex-Swann) [reply app/services/decorators/gross_income_summary_decorator.rb:43]: Agreed. I did this for this gross_income_collator and good to do this here too.

File: app/services/collators/gross_income_collator.rb [hunk 5]
```diff
@@ -1,22 +1,60 @@
 module Collators
   class GrossIncomeCollator < BaseWorkflowService
-    def call # rubocop:disable Metrics/MethodLength, Metrics/AbcSize
-      gross_income_summary.update!(
+    OPERATION = :credit
+    INCOME_CATEGORIES = CFEConstants::VALID_INCOME_CATEGORIES.map(&:to_sym)
+
+    def call
+      params = default_params
+
```
[the remark this answers is not in the document]
Comment (Alex-Swann) [reply app/services/collators/gross_income_collator.rb:8]: Yep, good point. There was a moment I thought I was working in a controller the other day and that is proof in the pudding in regards to that lol.

File: app/services/decorators/other_income_source_decorator.rb (hunk 1 above)
Comment (Alex-Swann) [reply app/services/decorators/other_income_source_decorator.rb:32]: I dipped my toe in the one liner water and the intention was found wanting. I thought there was a risk 😄 . 

With nomenclature, again I meant to update this, tricky as I was thinking I'm trying to find transactions of a particular type in the record but I know that clashes with the '#find_by' ActiveRecord standard. Would something like '#record_transaction_types` be more appropriate? It looks like naming conventions for methods prefixed with a verb tend to gear towards DB interactions, and anything else is more noun orientated. Is that sort of right out of interest? Open to any suggestions too.

File: app/services/collators/gross_income_collator.rb [hunk 6]
```diff
@@ -1,22 +1,60 @@
...
+      transactions = cash_transactions_find_by(operation: operation, name: name)
+      return 0.0 if transactions.empty?
+
+      transactions.average(:amount).round(2)
+    end
+
+    def cash_transactions_find_by(operation:, name:)
+      category = gross_income_summary.cash_transaction_categories.find_by(operation: operation, name: name)
+      CashTransaction.where(cash_transaction_category_id: category&.id)
+    end
```
[the remark this answers is not in the document]
Comment (Alex-Swann) [reply app/services/collators/gross_income_collator.rb:58]: Cool. Will take a look.

File: app/services/decorators/other_income_source_decorator.rb (hunk 1 above)
Comment (Alex-Swann) [reply app/services/decorators/other_income_source_decorator.rb:32]: Looking at it closer we are getting income transactions by type, so ended up with this
```
    def income_transactions(transaction_type:)
      income_transactions = {}

      INCOME_CATEGORIES.each do |category|
        income_transactions[category] = @record["#{category}_#{transaction_type}"]
      end

      income_transactions
    end
    ```
    let me know if that works for you

Comment (Alex-Swann) [reply app/services/decorators/other_income_source_decorator.rb:32]: there's also the possibility of sticking this in a concern and reusing it for disposable outgoings and having something like 
```
    def transactions(categories:, transaction_type:)
      transactions = {}

      categories.each do |category|
        transactions[category] = @record["#{category}_#{transaction_type}"]
      end

      transactions
    end
    ```
    but we can worry about that when doing the next stage maybe

File: app/services/collators/gross_income_collator.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (Alex-Swann) [reply app/services/collators/gross_income_collator.rb:58]: The CashTransaction model's records, unlike Apply, only contain amount and date information and it's the category to which they refer to (cash_transaction_category_id) which contains the operation. I've added some scopes to replicate what I've done above in both the CashTransaction and CashTransactionCategory models to account for this. Let me know if you're happy with that.

File: app/services/collators/gross_income_collator.rb [hunk 7]
```diff
@@ -1,22 +1,54 @@
...
+
+      case assessment.version
+      when CFEConstants::LATEST_ASSESSMENT_VERSION
+        populate_attrs_v3 attrs
+      else
+        populate_attrs_v2 attrs
+      end
+
+      gross_income_summary.send(:update!, attrs)
+    end
```
Comment (stephenrichards) [inline app/services/collators/gross_income_collator.rb:16]: I think `gross_income_summary.update!(attrs)` will work here.

File: app/services/decorators/other_income_source_decorator.rb (hunk 1 above)
Comment (stephenrichards) [reply app/services/decorators/other_income_source_decorator.rb:47]: Much much better.  There is an argument for making it a concern, but then that means you have to parameterise the INCOME_CATEGORIES  / OUTGOING_CATEGORIES, so may not be worth it.  I'll leave that up to you to decide when you come to to the outogings, if you pick that ticket up.

**Review by stephenrichards — APPROVED** [review]

**Pull request merged**